### PR TITLE
Feature/add more interfaces

### PR DESCRIPTION
**Pull request by takebene**
## Pull Request

- [x] I have checked for similar PRs.
- [x] I have read the [contributing guidelines](https://github.com/4Soft-de/vec-model/blob/develop/.github/CONTRIBUTING.md).

### Changes

- [x] Code
- [ ] Documentation
- [x] Other: JAXB Binding Declaration

### Description

- Adds the new interfaces `HasOccurrenceOrUsage` (vec113) and `HasOccurrenceOrUsages` (vec120) with a default method to get OccurrenceOrUsage(s) of a specific class. This is done separately for the `vec113` module and the `vec120` module due to having different return types.
- Adds a new interface `HasPropertyType`.
- Adds a new interface HasCustomProperties with a default method to get CustomProperty of a specific class.
- Extended existing `HasX` interfaces to have `getXWithType` methods. Also made them `@FunctionalInterface`s.
  - Removed a few navigation methods due to that.
- Moved util classes into the `vec-common` module.

### CODE REVIEW

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * THE SOFTWARE.
+ * =========================LICENSE_END==================================
+ */
+package com.foursoft.vecmodel.vec120;
+
+import java.util.List;
+
+public interface HasOccurrenceOrUsages {
+
+    List<VecOccurrenceOrUsage> getOccurrenceOrUsage();
```
Comment (DManstrator) [inline vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java:32]: Shouldn't that be `getOccurrenceOrUsages` (plural)?

File: vec-common/src/main/java/com/foursoft/vecmodel/common/HasSpecifications.java [hunk 2]
```diff
@@ -28,12 +28,27 @@
 import com.foursoft.vecmodel.common.util.DelegationUtils;
 
 import java.util.List;
+import java.util.Optional;
 
-public interface HasSpecifications<X> {
+public interface HasSpecifications<X extends HasIdentification> {
```
Comment (DManstrator) [inline vec-common/src/main/java/com/foursoft/vecmodel/common/HasSpecifications.java:33]: Those `HasX` classes could all be marked as a `@FunctionalInterface`.

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java (hunk 1 above)
Comment (takebene) [reply vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java:32]: not possible the method in the vec model is called like this

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/predicates/VecPredicates.java [hunk 3]
```diff
@@ -121,13 +120,6 @@ private VecPredicates() {
         return reference -> reference.getParentWireRole().getWireElementReferences().size() == 1;
     }
 
-    // TODO Use enum for SpecialPartType when available
-    public static Predicate<VecDocumentVersion> isAssembly() {
```
Comment (takebene) [inline vec120/src/main/java/com/foursoft/vecmodel/vec120/predicates/VecPredicates.java:125]: Special part types are too specific

File: vec-common/src/main/java/com/foursoft/vecmodel/common/HasSpecifications.java (hunk 2 above)
Comment (takebene) [reply vec-common/src/main/java/com/foursoft/vecmodel/common/HasSpecifications.java:34]: Good idea, you can do this in new branch.

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java (hunk 1 above)
Comment (DManstrator) [reply vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java:32]: Okay, that's bad but explains it. Perhaps documentation should be added to reflect that, else you could assume that list is supposed to only contain one element.

Will that get fixed with VEC 2.0.0?

File: vec-common/src/main/java/com/foursoft/vecmodel/common/HasSpecifications.java (hunk 2 above)
Comment (DManstrator) [reply vec-common/src/main/java/com/foursoft/vecmodel/common/HasSpecifications.java:34]: I don't see an issue not to do it in this branch. Why would you want a new PR for that?

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/navigations/SegmentNavs.java [hunk 4]
```diff
@@ -50,7 +50,6 @@ private SegmentNavs() {
         return segment -> segment.getParentTopologySpecification().getParentDocumentVersion().getDocumentNumber();
     }
 
-    // TODO Make the segmentLengthClassification an Open Enumeration
```
Comment (DManstrator) [inline vec120/src/main/java/com/foursoft/vecmodel/vec120/navigations/SegmentNavs.java:53]: Why was that ToDo removed?

File: vec120/src/main/resources/vec122/vec_1.2.2-ext.xjb [hunk 5]
```diff
@@ -44,5 +44,15 @@
...
+    
+    <jxb:bindings multiple="true" node="//xs:complexType[.//xs:element[@name='OccurrenceOrUsage' and @type='xs:IDREFS']]">
+        <inheritance:implements>com.foursoft.vecmodel.vec120.HasOccurrenceOrUsages</inheritance:implements>
+    </jxb:bindings>
+
+    <jxb:bindings multiple="true" node="//xs:complexType[.//xs:element[@name='CustomProperty' and @type='vec:CustomProperty']]">
+        <inheritance:implements>com.foursoft.vecmodel.vec120.HasCustomProperties</inheritance:implements>
+    </jxb:bindings>
+
+
```
Comment (DManstrator) [inline vec120/src/main/resources/vec122/vec_1.2.2-ext.xjb:57]: Why so many new lines?

File: vec-common/src/main/java/com/foursoft/vecmodel/common/HasRoles.java [hunk 6]
```diff
@@ -36,4 +38,18 @@
...
+     * Only use this method if you are sure there will just be one element!</b>
+     *
+     * @param type Class of T.
+     * @param <T>  Type of role to filter for.
+     * @return The first role with the given type if found, else an empty optional.
+     */
+    default <T extends X> Optional<T> getRoleWithType(final Class<T> type) {
+        return DelegationUtils.getFromListWithTypeAsStream(getRoles(), type)
+                .collect(StreamUtils.findOneOrNone());
+    }
```
[the remark this answers is not in the document]
Comment (takebene) [reply vec-common/src/main/java/com/foursoft/vecmodel/common/HasRoles.java:54]: i do not want to open a stream twice.

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/navigations/PartOccurrenceOrUsageNavs.java [hunk 7]
```diff
@@ -124,8 +124,7 @@ private PartOccurrenceOrUsageNavs() {
     // VecOccurrenceOrUsage
 
     public static BiFunction<VecOccurrenceOrUsage, VecDocumentVersion, Optional<VecTopologyNode>> findNodeOfComponent() {
-        return (component, documentVersion) -> component.getRolesWithType(VecPlaceableElementRole.class).stream()
-                .findFirst()  // there should be at most one such role
+        return (component, documentVersion) -> component.getRoleWithType(VecPlaceableElementRole.class)
```
[the remark this answers is not in the document]
Comment (takebene) [reply vec120/src/main/java/com/foursoft/vecmodel/vec120/navigations/PartOccurrenceOrUsageNavs.java:127]: that's what the method documentation is for

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/navigations/SegmentNavs.java (hunk 4 above)
Comment (takebene) [reply vec120/src/main/java/com/foursoft/vecmodel/vec120/navigations/SegmentNavs.java:53]: not sure if we're doing that in this project. I think also in such cases issues are better than todos

Comment (DManstrator) [reply vec120/src/main/java/com/foursoft/vecmodel/vec120/navigations/SegmentNavs.java:53]: Agree on that, I would open an issue then.

File: vec-common/src/main/java/com/foursoft/vecmodel/common/HasRoles.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (DManstrator) [reply vec-common/src/main/java/com/foursoft/vecmodel/common/HasRoles.java:54]: Fair point.

File: vec113/src/main/java/com/foursoft/vecmodel/vec113/HasCustomProperties.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package com.foursoft.vecmodel.vec113;
+
+import com.foursoft.vecmodel.common.util.DelegationUtils;
+import com.foursoft.vecmodel.common.util.StreamUtils;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface HasCustomProperties {
```
Comment (DManstrator) [inline vec113/src/main/java/com/foursoft/vecmodel/vec113/HasCustomProperties.java:34]: Shouldn't that class only exist once in the `vec-common` module?

File: vec113/src/main/java/com/foursoft/vecmodel/vec113/HasCustomProperties.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    List<VecCustomProperty> getCustomProperties();
+
+    default <T extends VecCustomProperty> List<T> getCustomProperties(final Class<T> type) {
+        return DelegationUtils.getFromListWithType(getCustomProperties(), type);
+    }
```
Comment (DManstrator) [inline vec113/src/main/java/com/foursoft/vecmodel/vec113/HasCustomProperties.java:40]: When migrated into one class, it should perhaps be `getCustomPropertiesWithType` to be consistent with the other interfaces. Forgot to mention that before.

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java (hunk 1 above)
Comment (takebene) [reply vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java:33]: Documation is generated via jaxb

File: vec-common/src/main/java/com/foursoft/vecmodel/common/util/StringUtils.java [hunk 10]
```diff
@@ -10,10 +10,10 @@
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
```
Comment (DManstrator) [inline vec-common/src/main/java/com/foursoft/vecmodel/common/util/StringUtils.java:10]: Still has `vec120` in the license.

File: vec113/src/main/java/com/foursoft/vecmodel/vec113/HasOccurrenceOrUsage.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
+/*-
+ * ========================LICENSE_START=================================
+ * vec-common
+ * %%
```
Comment (DManstrator) [inline vec113/src/main/java/com/foursoft/vecmodel/vec113/HasOccurrenceOrUsage.java:4]: Still has `vec-common` in the license.

File: vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
+/*-
+ * ========================LICENSE_START=================================
+ * vec-common
+ * %%
```
Comment (DManstrator) [inline vec120/src/main/java/com/foursoft/vecmodel/vec120/HasOccurrenceOrUsages.java:4]: Still has `vec-common` in the license.

**Review by DManstrator — DISMISSED** [review]

**Review by DManstrator — APPROVED** [review]

**Review by DManstrator — DISMISSED** [review]

**Review by DManstrator — APPROVED** [review]

**Pull request merged**